rcl/test/rcl/test_subscription.cpp

```diff
@@ -1741,3 +1757,124 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_init_fini_ma
     }
   });
 }
+
+struct TestParameters
+{
+  enum class TYPESUPPORT {C, CPP};
+
+  explicit TestParameters(
+    TYPESUPPORT pub_ts = TYPESUPPORT::C,
+    TYPESUPPORT sub_ts = TYPESUPPORT::CPP)
+  : pub_ts_(pub_ts), sub_ts_(sub_ts) {}
+
+  TYPESUPPORT pub_ts_;
+  TYPESUPPORT sub_ts_;
+};
+
+class TEST_FIXTURE_P_RMW (TestSubscriptionFixtureParam)
+  : public TEST_FIXTURE_P_RMW(TestSubscriptionFixture),
+  public ::testing::WithParamInterface<TestParameters>
+{
+protected:
+  void SetUp()
+  {
+    param_ = GetParam();
+    TEST_FIXTURE_P_RMW(TestSubscriptionFixture) ::SetUp();
+  }
+
+  TestParameters param_;
+};
+
+
+/* Test subscription to recevie complex message from a publisher with typesupport settings.
+ */
+TEST_P_RMW(TestSubscriptionFixtureParam, test_subscription_complex_message) {
+  rcl_ret_t ret;
+  const rosidl_message_type_support_t * ts_pub;
+  const rosidl_message_type_support_t * ts_sub;
+  if (param_.pub_ts_ == TestParameters::TYPESUPPORT::C) {
+    ts_pub = ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Arrays);
+  } else {
+    ts_pub = rosidl_typesupport_cpp::get_message_type_support_handle<test_msgs::msg::Arrays>();
+  }
+  if (param_.sub_ts_ == TestParameters::TYPESUPPORT::C) {
+    ts_sub = ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Arrays);
+  } else {
+    ts_sub = rosidl_typesupport_cpp::get_message_type_support_handle<test_msgs::msg::Arrays>();
+  }
+  constexpr char topic[] = "rcl_test_subscription_nominal_string_chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts_pub, topic, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+  rcl_subscription_t subscription = rcl_get_zero_initialized_subscription();
+  rcl_subscription_options_t subscription_options = rcl_subscription_get_default_options();
+  ret = rcl_subscription_init(&subscription, this->node_ptr, ts_sub, topic, &subscription_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_subscription_fini(&subscription, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+  ASSERT_TRUE(wait_for_established_subscription(&publisher, 10, 100));
+  constexpr char test_string[] = "testing";
+  constexpr bool bool_values[3] = {true, false, true};
+  {
```

Comment:
Any particular reason for this extra scope?

Reply:
To be honest, there's no particular reason. I just wanted to make the `rcl_publish / rcl_take` action inside a more clear block. 

I removed it in https://github.com/ros2/rcl/pull/1074/commits/d37e838d8a2f07c049838a36e891d39e854aa464.